app/Http/Controllers/Webapp/PostController.php

```diff
@@ -0,0 +1,1890 @@
+            $post->target_post_id = 0;
+            $this->_flashMessage($request, '更新に失敗しました。もう一度ご確認ください。' . $e->getMessage(), 'warning');
+        }
+
+        $post->save();
+        $post->saveHistory();
+
+        // Update Post meta and keywords
+        // Main keyword
+        PostMeta::deletePostMetaByPostId($post->id);
+        $main = explode(',', $request->get('mainkw'));
+
+        $uniqueMainKeyword = trim($main[0]);
+        $mainKeyword = Keywords::getKeyword($uniqueMainKeyword, 1);
+
+        if(empty($mainKeyword)) {//there is no keyword
+            if(strlen($uniqueMainKeyword) > 0) {
+                // Insert new main keyword
+                $mainKeyword = new Keywords();
+                $mainKeyword->keyword = $uniqueMainKeyword;
+                $mainKeyword->type = 1;
+                $mainKeyword->save();
+            }
+        }
+
+        $postMetaMkw = PostMeta::getPostMeta($post->id, isset($mainKeyword->id) ? $mainKeyword->id : 0);
+        if(empty($postMetaMkw) && isset($mainKeyword->id)) {
+            // Insert new post-keyword relationship if not exist
+            $postMetaMkw = new PostMeta();
+            $postMetaMkw->post_id = $post->id;
+            $postMetaMkw->keyword_id = $mainKeyword->id;
+            $postMetaMkw->save();
+        }
+
+        // Sub keywords
+        $subs = explode(',', $request->get('subkws'));
+        foreach ($subs as $item) {
+            $trimmedItem = trim($item);
+            if(strlen($trimmedItem) == 0) {
+                continue;
+            }
+
+            $subKeyword = Keywords::getKeyword($trimmedItem, 2);
+
+            if(empty($subKeyword)) {
+                // Insert new keyword
+                $subKeyword = new Keywords();
+                $subKeyword->keyword = $trimmedItem;
+                $subKeyword->type = 2;
+                $subKeyword->save();
+
+                // Save post-keyword relationship
+            }
+
+            $postMetaSkw = PostMeta::getPostMeta($post->id, $subKeyword->id);
+            if(empty($postMetaSkw)) {
+                // Insert new post-keyword relationship if not exist
+                $postMetaSkw = new PostMeta();
+                $postMetaSkw->post_id = $post->id;
+                $postMetaSkw->keyword_id = $subKeyword->id;
+                $postMetaSkw->save();
+            }
+
+
+        }
+
+        if(strpos($request->get('submit_type'),'公開') !== false || strpos($request->get('submit_type'),'下書き') !== false){
+            return redirect()->to(action('Webapp\\PostController@getEdit', $postId))->send();
+        }else if(strpos($request->get('submit_type'),'プレビュー') !== false){
+            return redirect()->to(action('Webapp\\PostController@getPreviewIframe', $postId))->send();
+        }
+
+        return redirect()->to(action('Webapp\\PostController@getEdit', $postId))->send();
+    }
+
+    public function postEditAjax(Request $request, $postId) {
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this function to reduce its Cognitive Complexity from 60 to the 15 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://sonar.relipa.blog/coding_rules#rule_key=php%3AS3776)
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Reduce the number of returns of this function 6, down to the maximum allowed 3. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://sonar.relipa.blog/coding_rules#rule_key=php%3AS1142)
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') This function "postEditAjax" has 182 lines, which is greater than the 150 lines authorized. Split it into smaller functions. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://sonar.relipa.blog/coding_rules#rule_key=php%3AS138)